File: Backend/src/main/java/com/chahd/collabproject/entity/Tache.java
package com.chahd.collabproject.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import org.hibernate.annotations.ColumnDefault;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@Entity
@ToString(exclude = {"affectations"})
@Table(name = "tache")
public class Tache {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    private Integer id;

    @NotNull
    @Column(name = "titre", nullable = false, length = Integer.MAX_VALUE)
    private String titre;

    @Column(name = "description", length = Integer.MAX_VALUE)
    private String description;

    @Size(max = 100)
    @NotNull
    @Column(name = "priorite", nullable = false, length = 100)
    private String priorite;

    @Column(name = "date_debut")
    private LocalDate dateDebut;

    @Column(name = "echeance")
    private LocalDate echeance;

    @Size(max = 100)
    @Column(name = "statut", length = 100)
    private String statut;

    @Column(name = "taux_avancement")
    private Double tauxAvancement;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "projet_id", nullable = false)
    private Projet projet;

    @NotNull
    @ColumnDefault("now()")
    @Column(name = "date_creation", nullable = false)
    private OffsetDateTime dateCreation;
    @OneToMany(mappedBy = "tache", fetch = FetchType.LAZY)
    @JsonIgnore
    private List<AffectationTache> affectations = new ArrayList<>();

    @Transient
    @JsonIgnore
    public List<Utilisateur> getCollaborateurs() {
        return affectations.stream()
                .filter(a -> a.getDateSuppression() == null)
                .map(AffectationTache::getCollaborateur)
                .toList();
    }

    @PrePersist
    protected void onCreate() {
        this.dateCreation = OffsetDateTime.now();
    }

}

File: Frontend/src/app/Model/tache.ts
import { Utilisateur } from "./Utilisateur";

export type StatutTache  = 'A_FAIRE' | 'EN_COURS' | 'TERMINEE'| 'BLOQUEE';
export type PrioriteTache  = 'HAUTE' | 'BASSE' | 'MOYENNE';


export class Tache {
  id!: number;
  titre!: string;
  description!: string;
  statut!: StatutTache;
  echeance!: string;
  projetId!: number;
  dateCreation!: string;
  priorite! :PrioriteTache ;
  dateDebut! : string;
   collaborateurs!: Utilisateur[];
   tauxAvancement !:Number;

  static fromJson(json: any): Tache {
    const tache = new Tache();
    tache.id = json.id;
    tache.titre = json.titre;
    tache.description = json.description;
    tache.statut = json.statut;
    tache.echeance = json.echeance;
    tache.dateDebut=json.dateDebut ;
    tache.projetId = json.projetId;
    tache.dateCreation = json.dateCreation;
    tache.priorite=json.priorite ;
     tache.collaborateurs = json.collaborateurs ?? [];
     tache.tauxAvancement=json.tauxAvancement ;
    return tache;
  }

 
}